lib/Minz/Migrator.php

```diff
@@ -0,0 +1,279 @@
+			// already at the latest version, so there is nothing more to do
+			return true;
+		}
+
+		$lock_path = $applied_migrations_path . '.lock';
+		if (!@mkdir($lock_path)) {
+			// Someone is probably already executing the migrations (the folder
+			// already exists).
+			// We should probably return something else, but we don't want the
+			// user to think there is an error (it's normal workflow), so let's
+			// stick to this solution for now.
+			// Another option would be to show him a maintenance page.
+			Minz_Log::warning(
+				'A request has been served while the application wasn’t up-to-date. '
+				. 'Too many of these errors probably means a previous migration failed.'
+			);
+			return true;
+		}
+
+		$results = $migrator->migrate();
+
+		foreach ($results as $migration => $result) {
+			if ($result === true) {
+				$result = 'OK';
+			} elseif ($result === false) {
+				$result = 'KO';
+			}
+
+			Minz_Log::notice("Migration {$migration}: {$result}");
+		}
+
+		$applied_versions = implode("\n", $migrator->appliedVersions());
+		$saved = file_put_contents($applied_migrations_path, $applied_versions);
+
+		if (!@rmdir($lock_path)) {
+			Minz_Log::error(
+				'We weren’t able to unlink the migration executing folder, '
+				. 'you might want to delete yourself: ' . $lock_path
+			);
+			// we don't return early because the migrations could have been
+			// applied successfully. This file is not "critical" if not removed
+			// and more errors will eventually appear in the logs.
+		}
+
+		if ($saved === false) {
+			return "Cannot save the {$applied_migrations_path} file";
+		}
+
+		if (!$migrator->upToDate()) {
+			// still not up to date? It means last migration failed.
+			return 'A migration failed to be applied, please see previous logs';
+		}
+
+		return true;
+	}
+
+	/**
+	 * Create a Minz_Migrator instance. If directory is given, it'll load the
+	 * migrations from it.
+	 *
+	 * All the files in the directory must declare a class named
+	 * <app_name>_Migration_<filename> with a static `migrate` method.
+	 *
+	 * - <app_name> is the application name declared in the APP_NAME constant
+	 * - <filename> is the migration file name, without the `.php` extension
+	 *
+	 * The files starting with a dot are ignored.
+	 *
+	 * @param string|null $directory
+	 *
+	 * @throws BadFunctionCallException if a callback isn't callable (i.e.
+	 *                                  cannot call a migrate method).
+	 */
+	public function __construct($directory = null)
+	{
+		$this->applied_versions = [];
+
+		if (!is_dir($directory)) {
+			return;
+		}
+
+		foreach (scandir($directory) as $filename) {
+			if ($filename[0] === '.') {
+				continue;
+			}
+
+			$filepath = $directory . '/' . $filename;
+			$migration_version = basename($filename, '.php');
+			$migration_class = APP_NAME . "_Migration_" . $migration_version;
+			$migration_callback = $migration_class . '::migrate';
+
+			$include_result = @include_once($filepath);
```

Comment:
It looks like we run this `foreach ... include` at each request, also when do not need a migration?

Reply:
It would be great to be able to return faster the information about the need to migrate or not.
Just a few quick ideas:
* Compare version in `constants.php` with a file `data/version.txt` and run through the migration code if they are not equal, then update `data/version.txt`
* Or provide a text file `available_migrations.txt` with the list of available migrations, to avoid having to scan a directory and include irrelevant code;
  * If timestamp of `applied_migrations.txt` is 0 (i.e. does not exist) or is older than `available_migrations.txt`, proceed.

What do you think @marienfressinaud ?

Reply:
I'll work on it this afternoon: the easiest solution that would require the less of work is to move this `include_once` call to the `migrate` function

Reply:
The easiest solution wasn't easy finally: it would make the whole class and tests more complicated.

I chose another solution: the versions names are extracted from the migration filenames, so I simply compare these names with the versions in the `applied_migrations.txt`.